### DISCUSSION TITLE
Put jwalk behind an interface and abstract away file system access

### DISCUSSION THREAD

**Discussion by gosuwachu** [Byron/dua-cli#180] [General]
fyi, I have started to work on abstracting away jwalk to hide it behind an interface so that it is easily swappable by a mock implementation that will allow writing unit tests for `Traversal::from_walk`. 

Branch with work in progress implementation can be found here: https://github.com/gosuwachu/dua-cli/tree/dev/pwach/testing-traversal

From my initial tests, there doesn't seem to be much if any impact on the performance.

**Comment by Byron** [comment]
I'd love to see the test-coverage improved for sure!

In theory, I'd think there will be no change in performance at all, and hope that the code won't be less readable due to the abstraction than it is now.

**Reply by gosuwachu** [reply]
My branch has version that is very much work in progress, but I think the readability will increase, as it should allow some simplification / deduplication, especially between interactive and aggregate mode. Also what has triggered me to work on this is I think there is a small bug there.

**Reply by gosuwachu** [reply]
fyi, I have worked on this for some time, and there was no bug. The logic was more complicated than I have thought. I have written a bunch of tests, but the code is quite ugly, so I am just going to leave it for now.

**Reply by Byron** [reply]
Thanks for the update!
